### ISSUE TITLE
feature request: Update internal dependents but don't release them

### ISSUE THREAD

**Issue by Blasz** [atlassian/changesets#641]
## Problem

There's a longstanding problem in monorepos & changesets where minimum versions of dependency version ranges can become stale and break consumers that still resolve dependencies to those stale versions.
For example,

* `pkg-a@1.0.0` depends on `pkg-b@^1.0.0`
* New API is added in a minor release to `pkg-b@1.1.0`
* `pkg-a@1.0.0` still depends on `pkg-b@^1.0.0` as it was not changed or released as part of `pkg-b`'s minor release
* After some time, `pkg-a` then consumes the new API of `pkg-b` and minor releases itself to `pkg-a@1.1.0`

`pkg-a`'s minimum version range of `pkg-b@^1.0.0` is now _stale_ as the minimum version actually needs to be `pkg-b@^1.1.0` instead. Any consumers that upgrade their `pkg-a` dependency from `1.0.0` to `1.1.0` via a standard `yarn add` will break due to the transitive `pkg-b` dependency still resolving to `1.0.0`.

The new experimental `updateInternalDependents` config option added in #542 addresses this issue by allowing internal dependents to always be bumped whenever a package is released. However, this approach is a bit heavy handed for our use-case in atlassian-frontend because:

1. It can cause a very large number of packages to be released for each small change due to the high interconnectedness between packages (and therefore large number of transitive dependents per package).
2. It can make it harder to incrementally upgrade certain packages without increasing the surface area to also upgrade all transitive dependencies

## Proposed solution

I propose two additional configuration capabilities:

1. The option to bump internal dependents but _not_ release them - similar to what occurs for `devDependencies`. This essentially defers the release of the dependents until the next time they are actively worked on - or until they are patch released due to a dependency version falling out of range. This would prevents the large number of packages released per changeset.
2. The option to only always bump internal dependents for minor releases or higher, rather than on patch releases as well. Following semver, new features/API should be added in minor releases rather than patch. So the problem of relying on newly added functionality of a dependency should only occur when the dependency is minor bumped.

**Comment by Andarist** [comment]
> It can make it harder to incrementally upgrade certain packages without increasing the surface area to also upgrade all transitive dependencies

Could you rephrase? I'm unsure if I got the intention right here - is it about transitive dependencies or transitive dependents?

> The option to bump internal dependents but not release them - similar to what occurs for devDependencies. This essentially defers the release of the dependents until the next time they are actively worked on - or until they are patch released due to a dependency version falling out of range. This would prevents the large number of packages released per changeset.

I think I see what you are saying here. If we reuse your example then we'd have this:
* `pkg-a@1.0.0` depends on `pkg-b@^1.0.0`
* New API is added in a minor release to `pkg-b@1.1.0`
* ~`pkg-a@1.0.0` still depends on `pkg-b@^1.0.0` as it was not changed or released as part of `pkg-b`'s minor release~ _the dependency range gets bumped in `pkg-a` to `pkg-b@^1.1.0` without actually releasing `pkg-a`_
* After some time, `pkg-a` then consumes the new API of `pkg-b` and minor releases itself to `pkg-a@1.1.0` _and everything is OK thanks to the adjusted previous step_

Could you confirm that this is what you are asking for?

> The option to only always bump internal dependents for minor releases or higher, rather than on patch releases as well. Following semver, new features/API should be added in minor releases rather than patch. So the problem of relying on newly added functionality of a dependency should only occur when the dependency is minor bumped.

That could probably be introduced as `"updateInternalDependents": "minor"` or similar. OTOH I wonder if we really need this if we'd implement your suggested change

**Comment by Blasz** [comment]
> > It can make it harder to incrementally upgrade certain packages without increasing the surface area to also upgrade all transitive dependencies
> 
> Could you rephrase? I'm unsure if I got the intention right here - is it about transitive dependencies or transitive dependents?

Dependencies. If you have multiple copies of a dependency in your node_modules, either due to a direct dependency + transitive or two separate transitive versions, you either have to de-duplicate them to a single version or incur the bundle size cost of having multiple versions.
If minimum version ranges keep getting bumped often, it reduces the set of versions to deduplicate to and increases the likelihood of having to upgrade existing versions of a dependency. Upgrading existing versions of a dependency solely for the purpose of deduplication increases the surface area of an upgrade as it was part of the original dependency you wished to upgrade.

An example,

* A consumer directly depends on `button@^1.0.0` and `modal-dialog@^1.0.0`. Modial-dialog depends on `button@^1.0.0`. In this case button is both a direct and transitive dependency has the chance of being duplicated by the consumer in the future.
* Consumer wishes to consume a new release of modal-dialog, say `modal-dialog@^1.1.0`. Before that new version of modal-dialog was released, button was patch released to `1.0.1` and modal-dialog's range on button was updated to `^1.0.1` via `"updateInternalDependents": "always"`
* Consumer now has to keep both versions of button around (1.0.0 and 1.0.1) or deduplicate them to 1.0.1 and increase the upgrade surface area from just instances of modal-dialog to instances of modal-dialog _and_ button.
* If modal-dialog's range on button was kept at `^1.0.0` then consumers would have the option of deduplicating button to 1.0.0 and limiting the surface area of the upgrade to modal-dialog only.

> > The option to bump internal dependents but not release them - similar to what occurs for devDependencies. This essentially defers the release of the dependents until the next time they are actively worked on - or until they are patch released due to a dependency version falling out of range. This would prevents the large number of packages released per changeset.
> 
> I think I see what you are saying here. If we reuse your example then we'd have this:
> 
> * `pkg-a@1.0.0` depends on `pkg-b@^1.0.0`
> * New API is added in a minor release to `pkg-b@1.1.0`
> * ~`pkg-a@1.0.0` still depends on `pkg-b@^1.0.0` as it was not changed or released as part of `pkg-b`'s minor release~ _the dependency range gets bumped in `pkg-a` to `pkg-b@^1.1.0` without actually releasing `pkg-a`_
> * After some time, `pkg-a` then consumes the new API of `pkg-b` and minor releases itself to `pkg-a@1.1.0` _and everything is OK thanks to the adjusted previous step_
> 
> Could you confirm that this is what you are asking for?

Yep that's right.

> > The option to only always bump internal dependents for minor releases or higher, rather than on patch releases as well. Following semver, new features/API should be added in minor releases rather than patch. So the problem of relying on newly added functionality of a dependency should only occur when the dependency is minor bumped.
> 
> That could probably be introduced as `"updateInternalDependents": "minor"` or similar. OTOH I wonder if we really need this if we'd implement your suggested change

They both do a good job of reducing the number of packages released for each change - although the deferred release capability is better in this regard. This is more aimed at allowing for more de-duplication on the consumer side.

**Comment by Andarist** [comment]
> If modal-dialog's range on button was kept at ^1.0.0 then consumers would have the option of deduplicating button to 1.0.0 and limiting the surface area of the upgrade to modal-dialog only.

Isn't the describing the opposite of what this feature request proposes? The issue mentioned that it would be desired for dependency ranges to be updated without creating releases for dependents and this text described kinda the opposite - that updated dependency range creates a bigger risk of duplication. The statement is true - it just seem to go into the opposite direction than this issue is proposing, so I kinda got confused here a little bit.
